### ISSUE TITLE
Omnifocus webDAV issues OC 7.0.4

### ISSUE THREAD

**Issue by jonathannz** [owncloud/core#13228]
Hello,
This may be related to issues:
https://github.com/owncloud/core/issues/1104 [this seems rather old to have any direct relevance, however the comment from "PreacherAT commented on Jun 3, 2013" exactly describes my issue]
and
https://github.com/owncloud/core/issues/10746

Reading through issue 10746, it is my understanding this was resolved in 7.0.3. So it may also be unrelated.

### Steps to reproduce
1. Configure OmniFocus to sync using WebDAV
2. Using the webdav link: https://mydomain.com/ocloud/remote.php/webdav/omnisync/
3. Try to initiate sync
4. Error occurs

### Expected behaviour
It should ultimately result in a folder called:
`OmniFocus.ofocus
Which will contain OmniFocus sync data files.

### Actual behaviour
Following error is reported:
>Unable to synchronize database with server.
>Unable to read document.
>No document exists at “https://mydomain.com/ocloud/remote.php/webdav/omnisync/OmniFocus.ofocus/”.

A folder named along the lines of this is created:
`OmniFocus.ofocus-copy-in-progress-lMjh1oxaW9C
There is nothing in the folder.

### Server configuration
**Operating system**:
Apache 2.2.26
**Web server:**
x86_64
**Database:**
mySQL 5.5.40-36.1
**PHP version:**
5.4.35
**ownCloud version:** (see ownCloud admin page)
7.0.4
**Updated from an older ownCloud or fresh install:**
Fresh
**List of activated apps:**
Activity
Cal
Con
Delete files
Docs
Encryption
External storage
First run
Full text search
Mail template
PDF viewer
Pictures
Share files
Text editor
Updater
Versions

**The content of config/config.php:**
```
<?php
$CONFIG = array (
  'instanceid' => 'oc2c32b5b288',
  'passwordsalt' ⇒ 'REMOVED',
  'secret' ⇒ 'REMOVED',
  'trusted_domains' => 
  array (
    0 ⇒ 'MYDOMAIN.COM',
  ),
  'datadirectory' => '/home/inspired/oclouddata',
  'overwrite.cli.url' => 'http://MYDOMAIN.com/ocloud',
  'dbtype' => 'mysql',
  'version' => '7.0.4.2',
  'dbname' => 'inspired_ocloud1',
  'dbhost' => 'localhost',
  'dbtableprefix' => 'oc_',
  'dbuser' => 'inspired_ocloudu',
  'dbpassword' ⇒ 'REMOVED',
  'installed' => true,
  'forcessl' => true,
);
```

**Are you using external storage, if yes which one:** local/smb/sftp/...
NO
**Are you using encryption:** yes/no
YES
### Client configuration
Issue is not browser dependant.

#### Web server error log
```
Insert your webserver log here
```

#### ownCloud log (data/owncloud.log)
```
{"app":"Encryption library","message":"No share key found for user \"jonathan\" for file \"OmniFocus.ofocus-copy-in-progress-kdavwbdq4Ju\"","level":2,"time":"2015-01-09T22:53:49+00:00"}
{"app":"Encryption library","message":"No share key found for user \"jonathan\" for file \"OmniFocus.ofocus\"","level":2,"time":"2015-01-09T22:54:47+00:00"}
{"app":"PHP","message":"Array to string conversion at \/home\/inspired\/public_html\/ocloud\/lib\/private\/l10n\/string.php#57","level":3,"time":"2015-01-09T23:56:39+00:00"}
{"app":"Encryption library","message":"No share key found for user \"jonathan\" for file \"OmniFocus.ofocus-copy-in-progress-crfbg7gpmy_\"","level":2,"time":"2015-01-10T00:12:11+00:00"}
{"app":"Encryption library","message":"No share key found for user \"jonathan\" for file \"OmniFocus.ofocus-copy-in-progress-lMjh1oxaW9C\"","level":2,"time":"2015-01-10T00:12:11+00:00"}
{"app":"core","message":"App does not provide an info.xml file","level":3,"time":"2015-01-10T00:20:49+00:00"}
```

**Comment by jonathannz** [comment]
I have disabled Encryption, and the issue persists. So it does not appear to be an encryption related problem.

**Comment by PVince81** [comment]
Can you post the access log so we can see what omnifocus is trying to do over WebDAV ?

**Comment by jonathannz** [comment]
I am not sure where to retrieve an "access log" from. Please advise.
Here's data from the log found in the Admin section of OC.
```
Debug	search_lucene	adding /OmniFocus.ofocus-copy-in-progress-bqViPbgIcq9	2015-01-15T02:15:31+00:00
Debug	search_lucene	found 0 hits	2015-01-15T02:15:31+00:00
Debug	search_lucene	searching hits for pk:2153feee9427048e3822cf3ea68ed579	2015-01-15T02:15:31+00:00
Debug	webdav	Sabre\DAV\Exception\NotFound: File with name OmniFocus.ofocus-copy-in-progress-aGQiP8wUwaD could not be located	2015-01-15T02:14:55+00:00
Debug	webdav	Sabre\DAV\Exception\NotFound: File with name OmniFocus.ofocus could not be located	2015-01-15T02:14:53+00:00
Debug	webdav	Sabre\DAV\Exception\NotAuthenticated: No basic authentication headers were found	2015-01-15T02:14:51+00:00
Debug	webdav	Sabre\DAV\Exception\NotFound: File with name OmniFocus.ofocus could not be located	2015-01-15T02:14:49+00:00
Debug	webdav	Sabre\DAV\Exception\NotAuthenticated: No basic authentication headers were found	2015-01-15T02:14:48+00:00
Debug	webdav	Sabre\DAV\Exception\NotFound: File with name OmniFocus.ofocus could not be located	2015-01-15T02:14:03+00:00
Debug	webdav	Sabre\DAV\Exception\NotFound: File with name OmniFocus.ofocus-copy-in-progress-bqViPbgIcq9 could not be located	2015-01-15T02:13:27+00:00
Debug	webdav	Sabre\DAV\Exception\NotFound: File with name OmniFocus.ofocus could not be located	2015-01-15T02:13:26+00:00
Debug	webdav	Sabre\DAV\Exception\NotFound: File with name OmniFocus.ofocus could not be located	2015-01-15T02:13:24+00:00
Debug	webdav	Sabre\DAV\Exception\NotFound: File with name OmniFocus.ofocus could not be located	2015-01-15T02:12:47+00:00
Debug	webdav	Sabre\DAV\Exception\NotAuthenticated: No basic authentication headers were found	2015-01-15T02:12:46+00:00
Debug	carddav	Sabre\DAV\Exception\NotAuthenticated: No basic authentication headers were found	2015-01-15T00:40:33+00:00
Debug	carddav	Sabre\DAV\Exception\NotAuthenticated: No basic authentication headers were found	2015-01-15T00:40:29+00:00
Debug	carddav	Sabre\DAV\Exception\NotAuthenticated: No basic authentication headers were found	2015-01-15T00:30:53+00:00
Debug	carddav	Sabre\DAV\Exception\NotAuthenticated: No basic authentication headers were found	2015-01-15T00:30:50+00:00
Debug	carddav	Sabre\DAV\Exception\NotAuthenticated: No basic authentication headers were found	2015-01-15T00:27:55+00:00
Debug	carddav	Sabre\DAV\Exception\NotAuthenticated: No basic authentication headers were found	2015-01-15T00:27:52+00:00
Debug	carddav	Sabre\DAV\Exception\NotAuthenticated: No basic authentication headers were found	2015-01-15T00:23:15+00:00
Debug	carddav	Sabre\DAV\Exception\NotAuthenticated: No basic authentication headers were found	2015-01-15T00:23:12+00:00
Debug	carddav	Sabre\DAV\Exception\NotAuthenticated: No basic authentication headers were found
```

**Comment by PVince81** [comment]
You might find the access log in `/var/log/apache2/access_log`. It might also be called `access.log`.

Once you find it, please re-run the sync scenario to make it produce results, then post it here :smile:

**Comment by jonathannz** [comment]
This is on a shared server environment, so as far as I know I have no access to /var
There’s a LOT of data being generated in the owncloud.log file I posted some lines from earlier.

But most a repeat of the previous data, and a great deal more along the lines of:

{"reqId":"54b3049e9521e","app":"carddav","message":"Sabre\\DAV\\Exception\\NotAuthenticated: No basic authentication headers were found","level":0,"time":"2015-01-11T23:17:50+00:00","method":"PROPFIND","url":"\/ocloud\/remote.php\/carddav\/addressbooks\/jonathan\/"}

and 

{"reqId":"54b859e216721","app":"carddav","message":"Sabre\\DAV\\Exception\\NotAuthenticated: No basic authentication headers were found","level":0,"time":"2015-01-16T00:22:58+00:00","method":"OPTIONS","url":"\/ocloud\/remote.php\/carddav\/principals\/jonathan\/"}

But I gather these are unrelated, as they look like Carddav issues.

**Comment by PVince81** [comment]
Is this still happening with 7.0.7 or 8.0.5 ?

**Comment by PVince81** [comment]
The access log could help a lot to find out what exactly Omnifocus is trying to do with Webdav that causes issues.

**Comment by Zettt** [comment]
Hello @PVince81,

> Do you mind adding your setup to this thread, mostly which ownCloud
version you are using and whether encryption or the external storage
apps are used?

I'm using 8.1.0 (stable). For the moment encryption is off, and we're not using external storage apps. The files are hosted on Uberspace.

> Is this about files only or also calendar and contacts?

Just files.

> Do you have a snippet of the access log around the error?

Yes, this is OmniPresence:

```
The xxx.com server returned “precondition failed” (412) in response to a request to “MOVE /remote.php/webdav/OmniDAV-Conformance-Tests-ahHK6PhYMBy/fortestReplacedCollectionUpdatesModificationDate: /parent/tmp/doc”.
```

`tail owncloud.log`: 

```
{"reqId":"GGS+kz3rU4Xv+L8qvs6K","remoteAddr":"46.223.86.111","app":"webdav","message":"Exception: {\"Message\":\"HTTP\\\/1.1 409 The destination node is not found\",\"Code\":0,\"Trace\":\"#0 \\\/var\\\/www\\\/virtual\\\/me\\\/html\\\/xxx.com\\\/3rdparty\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/CorePlugin.php(617): Sabre\\\\DAV\\\\Server->getCopyAndMoveInfo(Object(Sabre\\\\HTTP\\\\Request))\\n#1 [internal function]: Sabre\\\\DAV\\\\CorePlugin->httpMove(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#2 \\\/var\\\/www\\\/virtual\\\/me\\\/html\\\/xxx.com\\\/3rdparty\\\/sabre\\\/event\\\/lib\\\/EventEmitterTrait.php(105): call_user_func_array(Array, Array)\\n#3 \\\/var\\\/www\\\/virtual\\\/me\\\/html\\\/xxx.com\\\/3rdparty\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(469): Sabre\\\\Event\\\\EventEmitter->emit('method:MOVE', Array)\\n#4 \\\/var\\\/www\\\/virtual\\\/me\\\/html\\\/xxx.com\\\/3rdparty\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(254): Sabre\\\\DAV\\\\Server->invokeMethod(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#5 \\\/var\\\/www\\\/virtual\\\/me\\\/html\\\/xxx.com\\\/apps\\\/files\\\/appinfo\\\/remote.php(83): Sabre\\\\DAV\\\\Server->exec()\\n#6 \\\/var\\\/www\\\/virtual\\\/me\\\/html\\\/xxx.com\\\/remote.php(132): require_once('\\\/var\\\/www\\\/virtua...')\\n#7 {main}\",\"File\":\"\\\/var\\\/www\\\/virtual\\\/me\\\/html\\\/xxx.com\\\/3rdparty\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php\",\"Line\":746}","level":4,"time":"2015-08-10T14:29:00+00:00"}
{"reqId":"qNEjJibwvv4TBvJQSJa1","remoteAddr":"46.223.86.111","app":"webdav","message":"Exception: {\"Message\":\"HTTP\\\/1.1 409 The destination node is not found\",\"Code\":0,\"Trace\":\"#0 \\\/var\\\/www\\\/virtual\\\/me\\\/html\\\/xxx.com\\\/3rdparty\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/CorePlugin.php(617): Sabre\\\\DAV\\\\Server->getCopyAndMoveInfo(Object(Sabre\\\\HTTP\\\\Request))\\n#1 [internal function]: Sabre\\\\DAV\\\\CorePlugin->httpMove(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#2 \\\/var\\\/www\\\/virtual\\\/me\\\/html\\\/xxx.com\\\/3rdparty\\\/sabre\\\/event\\\/lib\\\/EventEmitterTrait.php(105): call_user_func_array(Array, Array)\\n#3 \\\/var\\\/www\\\/virtual\\\/me\\\/html\\\/xxx.com\\\/3rdparty\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(469): Sabre\\\\Event\\\\EventEmitter->emit('method:MOVE', Array)\\n#4 \\\/var\\\/www\\\/virtual\\\/me\\\/html\\\/xxx.com\\\/3rdparty\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(254): Sabre\\\\DAV\\\\Server->invokeMethod(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#5 \\\/var\\\/www\\\/virtual\\\/me\\\/html\\\/xxx.com\\\/apps\\\/files\\\/appinfo\\\/remote.php(83): Sabre\\\\DAV\\\\Server->exec()\\n#6 \\\/var\\\/www\\\/virtual\\\/me\\\/html\\\/xxx.com\\\/remote.php(132): require_once('\\\/var\\\/www\\\/virtua...')\\n#7 {main}\",\"File\":\"\\\/var\\\/www\\\/virtual\\\/me\\\/html\\\/xxx.com\\\/3rdparty\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php\",\"Line\":746}","level":4,"time":"2015-08-10T14:29:01+00:00"}
{"reqId":"1pd3\/0E1Kl8eu9Jvhv55","remoteAddr":"46.223.86.111","app":"webdav","message":"Exception: {\"Message\":\"HTTP\\\/1.1 409 The destination node is not found\",\"Code\":0,\"Trace\":\"#0 \\\/var\\\/www\\\/virtual\\\/me\\\/html\\\/xxx.com\\\/3rdparty\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/CorePlugin.php(617): Sabre\\\\DAV\\\\Server->getCopyAndMoveInfo(Object(Sabre\\\\HTTP\\\\Request))\\n#1 [internal function]: Sabre\\\\DAV\\\\CorePlugin->httpMove(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#2 \\\/var\\\/www\\\/virtual\\\/me\\\/html\\\/xxx.com\\\/3rdparty\\\/sabre\\\/event\\\/lib\\\/EventEmitterTrait.php(105): call_user_func_array(Array, Array)\\n#3 \\\/var\\\/www\\\/virtual\\\/me\\\/html\\\/xxx.com\\\/3rdparty\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(469): Sabre\\\\Event\\\\EventEmitter->emit('method:MOVE', Array)\\n#4 \\\/var\\\/www\\\/virtual\\\/me\\\/html\\\/xxx.com\\\/3rdparty\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(254): Sabre\\\\DAV\\\\Server->invokeMethod(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#5 \\\/var\\\/www\\\/virtual\\\/me\\\/html\\\/xxx.com\\\/apps\\\/files\\\/appinfo\\\/remote.php(83): Sabre\\\\DAV\\\\Server->exec()\\n#6 \\\/var\\\/www\\\/virtual\\\/me\\\/html\\\/xxx.com\\\/remote.php(132): require_once('\\\/var\\\/www\\\/virtua...')\\n#7 {main}\",\"File\":\"\\\/var\\\/www\\\/virtual\\\/me\\\/html\\\/xxx.com\\\/3rdparty\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php\",\"Line\":746}","level":4,"time":"2015-08-10T14:29:01+00:00"}
```



In ownCloud 8.1 we upgraded the SabreDAV library which is used for
Webdav, hopefully this might help fixing the problem ?

**Comment by PVince81** [comment]
Hmmm that could be the reason:
`MOVE /remote.php/webdav/OmniDAV-Conformance-Tests-ahHK6PhYMBy/fortestReplacedCollectionUpdatesModificationDate: /parent/tmp/doc`

If I understand well it's trying to move a folder into "/parent/tmp/doc" instead of "/remote.php/webdav/parent/tmp/doc".
Could there be a problem with root folder handling ?

@Zettt was "/parent/tmp/doc" a folder that you explicitly created ? Did you ask Omnifocus to move stuff there ?

Or does Omnifocus expect to have access to something in the root of the URL ? But this would mean that Omnifocus would be unable to work with any DAV server (not only ownCloud) running under a subpath like "http://example.com/davserverroothere/"

**Comment by Zettt** [comment]
As mentioned the Omni app in question here is [OmniPresence](https://www.omnigroup.com/omnipresence. OmniFocus probably uses the same implementation and has the same issues though, since it would make most sense to me that Omni makes one WebDAV framework that it ships with all their apps. OmniPresence is testing the server and in doing so is testing the `MOVE` command. Is that helpful?

**Comment by PVince81** [comment]
It would be good if I could get more details about the MOVE command, especially what "Destination" header it is sending. There is a newew Webdav standard that allow the "Destination" header to be a relative path instead of absolute path (absolute as in "pass the full URL"). I'm not sure whether SabreDAV supports relative URLs. I can test that with curl commands.

**Comment by PVince81** [comment]
These work:
- Absolute target: `curl -X MOVE -u root:admin "http://localhost/owncloud/remote.php/webdav/test.txt" -H 'Destination: http://localhost/owncloud/remote.php/webdav/test-absolute.txt'`
- Target relative to "owncloud": `curl -X MOVE -u root:admin "http://localhost/owncloud/remote.php/webdav/test.txt" -H 'Destination: /owncloud/remote.php/webdav/test-relative.txt'`

This doesn't:
```
% curl -X MOVE -u root:admin "http://localhost/owncloud/remote.php/webdav/test.txt" -H 'Destination: test-relative2.txt'
<?xml version="1.0" encoding="utf-8"?>
<d:error xmlns:d="DAV:" xmlns:s="http://sabredav.org/ns">
  <s:exception>Sabre\DAV\Exception\Forbidden</s:exception>
  <s:message>Requested uri (test-relative2.txt) is out of base uri (/owncloud/remote.php/webdav/)</s:message>
</d:error>
```

But here I see that the exception is 403 forbidden.

The exception above was actually 409 Conflict.

This can be reproduced by trying to move to a non-existing folder:
```
  % curl -X MOVE -u root:admin "http://localhost/owncloud/remote.php/webdav/test.txt" -H 'Destination: http://localhost/owncloud/remote.php/webdav/unexist/folder'
<?xml version="1.0" encoding="utf-8"?>
<d:error xmlns:d="DAV:" xmlns:s="http://sabredav.org/ns">
  <s:exception>Sabre\DAV\Exception\Conflict</s:exception>
  <s:message>The destination node is not found</s:message>
</d:error>
```

So this means that Omni* is trying to move into a non-existing folder.

Can you check whether the folder "http://example.com/owncloudpath/remote.php/webdav/parent/tmp/doc" actually exists ? Is Omni* trying to do several things in parallel and creates the folder without waiting for it to exist ?

**Comment by PVince81** [comment]
(this is all guesswork, I don't have access to any Omni* instance to test this myself)

**Comment by mvrueden** [comment]
I had the same issue and was not able to get `OmniFocus` to sync with my owncloud instance. 
I then saw, that the URL I set was something like `https://example.com/owncloud/remote.php/webdav/` (see the `/` at the end). 
This somehow looked wrong to me. After i removed the trailing `/` everything works fine, though I am not completely sure if it is related in any way.

**Comment by brevilo** [comment]
> After i removed the trailing / everything works fine

Unfortunately this workaround doesn't apply to OmniPresence 1.4.1 as it always appends a trailing slash to the server URL such that the error described in https://github.com/owncloud/core/issues/13228#issuecomment-129476117 still persists (OC 8.2.1).

**Comment by PVince81** [comment]
Can someone capture the Webdav request, including headers ?
Would be good to have an equivalent `curl` command to be able to reproduce locally.

**Comment by LukasReschke** [comment]
I'm pretty confident that this is a duplicate of https://github.com/owncloud/core/issues/8033

**Comment by PVince81** [comment]
Okay, closing in favor of https://github.com/owncloud/core/issues/8033#issuecomment-48684115

See there is more information, something about etags and "If" headers.
